src/firefly/java/edu/caltech/ipac/util/download/RetrieveUtil.java

```diff
@@ -0,0 +1,223 @@
+        String outLoc = resType==ResourceType.S3 ?  loc : FileUtil.makeShortHostName(loc);
+        String prefix= resType==ResourceType.S3 ? "S3-" : "URL-";
+        retval = prefix + outLoc + "-" + originalHashCode + baseKey;
+        //note: "=","," signs causes problem in download servlet.
+        retval = retval.replaceAll("[ :\\[\\]/\\\\|*?<>=,]", "-");
+        String ext= FileUtil.getExtension(fileStr,true);
+        var fileExt= DEF_EXT_LIST.contains(ext) ? ext : DEF_EXT_LIST.getFirst();
+        return retval + "." + fileExt;
+    }
+
+    public static AnyUrlParams makeFromURIString(String uriStr) throws FailedRequestException {
+
+        if (uriStr == null) throw new FailedRequestException("URL is null");
+        uriStr= uriStr.trim();
+        if (S3Download.S3Ref.isS3Ref(uriStr)) {
+            S3Download.S3Ref ref= S3Download.S3Ref.makeFromUri(uriStr);
+            return new AnyUrlParams(null,ref,uriStr,null);
+        }
+        else {
+            try {
+                return new AnyUrlParams(new URI(uriStr).toURL(), uriStr,null);
+            } catch (URISyntaxException | MalformedURLException e) {
+                throw new FailedRequestException("bad url",e);
+            }
+        }
+    }
+
+    public static ResourceType getResourceType(Object obj)  {
+        if (obj == null) return null;
+        if (S3Download.S3Ref.isS3Ref(obj)) {
+            return ResourceType.S3;
+        }
+        else if (obj instanceof URL) {
+            return ResourceType.URL;
+        }
+        else if (obj instanceof String s ){
+            try {
+                var ignore= new URI(s.trim()).toURL();
+                return ResourceType.URL;
+            } catch (URISyntaxException | MalformedURLException ignore) {
+                return null;
+            }
+        }
+        else {
+            return null;
+        }
+    }
+
+
+    /**
+     * if both parameters are non-null then we are assuming they are referencing the same piece of data
+     * @param ref the s3 reference of a piece of data
+     * @param urlStr the url reference to a piece of data
+     * @return the optimal way to retrieve the data
+     */
+    public static ResourceType detemineOptimalResourceType(S3Download.S3Ref ref, String urlStr) {
+        if (ref==null && urlStr==null) return null;
+
+        if (getResourceType(urlStr)!=ResourceType.URL) return ResourceType.S3;
+        String serverHost= null;
+        String resHost= null;
+        try {
+            serverHost= new URI(ServerContext.getRequestOwner().getBaseUrl()).toURL().getHost();
+            resHost= new URI(urlStr).toURL().getHost();
+        } catch (URISyntaxException | MalformedURLException ignore) { }
+        if (ref==null) return ResourceType.URL;
+        if (serverHost==null || resHost==null) return ResourceType.S3;
+        return FileUtil.makeHostDomain(serverHost).equals(FileUtil.makeHostDomain(resHost))
+                ? ResourceType.URL : ResourceType.S3;
```

Comment:
My interpretation is that `cloud_access` should be used when running in a cloud environment; otherwise, use `url_access`.  
One approach is to introduce a property (e.g., runtime_platform) and use it when available, falling back to auto-detection if it’s not set. There are ways to detect whether you’re running in AWS or GCP. This is also useful when both URIs are provided in `cloud_access`.